File: TunnelProject/Opc/OpcObjectDictionary.cs
namespace TunnelProject.Opc
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using TunnelProject.Opc.Interfaces;

    internal class OpcObjectDictionary : Dictionary<OPCObjectKey, IOpcObject>
    {
        public bool Contains(string plcName, string tagName)
        {
            foreach (OPCObjectKey k in Keys)
            {
                if (k.TagName == tagName && k.PLCName == plcName)
                {
                    return true;
                }
            }
            return false;
        }

        public void Remove(string groupName)
        {
            var itemsToBeRemoved = new List<OPCObjectKey>();
            foreach (OPCObjectKey key in Keys)
            {
                if (key.GroupName == groupName)
                {
                    itemsToBeRemoved.Add(key);
                }
            }
            foreach (OPCObjectKey key in itemsToBeRemoved)
            {
                this[key] = null;
                base.Remove(key);
            }
        }

        public IOpcObject Item(string plcName, string tagName)
        {
            foreach (OPCObjectKey s in Keys.ToList())
            {
                if (s.PLCName == plcName && s.TagName == tagName)
                {
                    return base[s];
                }
            }
            throw new ArgumentException("Key Does Not Exist");
        }

        public IOpcObject Item(OPCObjectKey key)
        {
            return base[key];
        }
    }
}

File: TSA_Rail_System_TunnelProject/Wdw_Sust.OpcDa.Core/OpcObjectDictionary.vb
Imports Wdw_Sust.OpcDa.Core.Interfaces

Friend Class OpcObjectDictionary
    Inherits Dictionary(Of OpcObjectKey, IOpcObject)

    Public Shadows Function Contains(plcName As String, tagName As String) As Boolean
        For Each k As OpcObjectKey In Keys
            If k.TagName = tagName AndAlso k.PLCName = plcName Then
                Return True
            End If
        Next
        Return False
    End Function

    Public Shadows Function Item(plcName As String, tagName As String) As IOpcObject
        For Each s As OpcObjectKey In Keys.ToList
            If s.PLCName = plcName AndAlso s.TagName = tagName Then
                Return MyBase.Item(s)
            End If
        Next
        Throw New ArgumentException("Key Does Not Exist")
    End Function

    Public Shadows Function Item(key As OpcObjectKey) As IOpcObject
        Return MyBase.Item(key)
        Throw New ArgumentException("Key Does Not Exist")
    End Function

    Public Overloads Sub Remove(groupName As String)
        Dim itemsToBeRemoved As New List(Of OpcObjectKey)
        For Each key As OpcObjectKey In Keys
            If key.GroupName = groupName Then
                itemsToBeRemoved.Add(key)
            End If
        Next
        For Each key As OpcObjectKey In itemsToBeRemoved
            Me(key) = Nothing
            MyBase.Remove(key)
        Next
    End Sub

End Class